internal/pkg/cli/deploy/workload.go

```diff
@@ -439,36 +427,36 @@ type UploadArtifactsOutput struct {
 	CustomResourceURLs map[string]string
 }
 
-func (d *workloadDeployer) uploadArtifacts() (*UploadArtifactsOutput, error) {
-	imageDigests, err := d.uploadContainerImages(d.imageBuilderPusher)
-	if err != nil {
-		return nil, err
-	}
-	s3Artifacts, err := d.uploadArtifactsToS3(&uploadArtifactsToS3Input{
-		fs:       d.fs,
-		uploader: d.s3Client,
-	})
-	if err != nil {
-		return nil, err
-	}
+// uploadArtifactFunc uploads an artifact and updates out
+// with any relevant information to be returned by uploadArtifacts.
+type uploadArtifactFunc func(out *UploadArtifactsOutput) error
 
-	out := &UploadArtifactsOutput{
-		ImageDigests: imageDigests,
-		EnvFileARNs:  s3Artifacts.envFileARNs,
-		AddonsURL:    s3Artifacts.addonsURL,
+// uploadArtifacts runs each of the uploadArtifact functions sequentially and returns
+// the output built by each of those functions. It short-circuts and returns
+// the error if one of steps returns an error.
```

Comment:
Would the order ever matter? If an error IS returned, does the user need to know what got uploaded and what didn't?

Reply:
right now, i don't think it matters what order each of the uploads happens in? theoretically, we could probably do the uploads in parallel if we wanted. 

as for the uploads - i think it could be nice for the static site to let them know what gets uploaded. however - that actually happens somewhere else lol: https://github.com/aws/copilot-cli/blob/12d5060828892efd017d69fdf5a1885aacfdaf6e/internal/pkg/cli/deploy/static_site.go#L90-L102